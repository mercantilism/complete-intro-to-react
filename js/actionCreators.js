// @flow

import { SET_SEARCH_TERM, ADD_API_DATA } from './actions';

export function setSearchTerm(searchTerm: string) {
  return { type: SET_SEARCH_TERM, payload: searchTerm };
}

// We're not actually going to be using this outside this file,
// but it's a good idea to export it anyway, so that we can import it
// into our test suite, or reuse on a separate incoming data source we might have
export function addAPIData(apiData: Show) {
  return { type: ADD_API_DATA, payload: apiData };
}
