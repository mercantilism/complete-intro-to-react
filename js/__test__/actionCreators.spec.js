// @flow

// moxios is the mocking library for axios
import moxios from 'moxios';
import { setSearchTerm, addAPIData, getAPIDetails } from '../actionCreators';

const westWorld = {
  rating: '6.0',
  title: 'Westworld',
  year: '2016â€“',
  description:
    'Set at the intersection of the near future and the reimagined past, explore a world in which every human appetite, no matter how noble or depraved, can be indulged without consequence.',
  poster: 'ww.jpg',
  imdbID: 'tt0475784',
  trailer: 'eX3u0IlBBO4'
};

test('setSearchTerm', () => {
  expect(setSearchTerm('New York')).toMatchSnapshot();
});

test('addAPIData', () => {
  expect(addAPIData(westWorld)).toMatchSnapshot();
});

// Note that the test for getAPIData uses addAPIData internally, and so is predicated on a good test for the addAPIData

// Done is here specific to jest and used for async tests, done will trigger the test completeion,
// otherwise the test will not complete until it has encounterd don
test('getAPIDetails', (done: Function) => {
  // this substitutes the mock version of axios instead of actually making a reuest
  const dispatchMock = jest.fn();
  moxios.withMock(() => {
    // getAPIDetails returns the thunk, a function normally then called by dispatch, here called by dispatchMock
    getAPIDetails(westWorld.imdbID)(dispatchMock);
    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request
        .respondWith({
          status: 200,
          response: westWorld
        })
        .then(() => {
          expect(request.url).toEqual(`http://localhost:3000/${westWorld.imdbID}`);
          expect(dispatchMock).toBeCalledWith(addAPIData(westWorld));
          done();
        });
    });
  });
});
