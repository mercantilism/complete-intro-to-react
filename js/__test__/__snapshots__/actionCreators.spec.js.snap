// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addAPIData 1`] = `
Object {
  "payload": Object {
    "description": "Set at the intersection of the near future and the reimagined past, explore a world in which every human appetite, no matter how noble or depraved, can be indulged without consequence.",
    "imdbID": "tt0475784",
    "poster": "ww.jpg",
    "rating": "6.0",
    "title": "Westworld",
    "trailer": "eX3u0IlBBO4",
    "year": "2016â€“",
  },
  "type": "ADD_API_DATA",
}
`;

exports[`setSearchTerm 1`] = `
Object {
  "payload": "New York",
  "type": "SET_SEARCH_TERM",
}
`;
